osf/models/preprint_provider.py

```diff
@@ -34,24 +33,34 @@ def __unicode__(self):
 
     @property
     def top_level_subjects(self):
-        if len(self.subjects_acceptable) == 0:
-            return Subject.find(Q('parents', 'isnull', True))
-        tops = set([sub[0][0] for sub in self.subjects_acceptable])
-        return [Subject.load(sub) for sub in tops]
+        if self.subjects.exists():
+            return self.subjects.filter(parent__isnull=True)
+        else:
+            # TODO: Delet this when all PreprintProviders have a mapping
+            from osf.models.subject import Subject
+            if len(self.subjects_acceptable) == 0:
+                return Subject.find(Q('parent', 'isnull', True))
+            tops = set([sub[0][0] for sub in self.subjects_acceptable])
+            return [Subject.load(sub) for sub in tops]
 
     @property
     def all_subjects(self):
-        q = []
-        for rule in self.subjects_acceptable:
-            if rule[1]:
-                q.append(Q('parents', 'eq', Subject.load(rule[0][-1])))
-                if len(rule[0]) == 1:
-                    potential_parents = Subject.find(Q('parents', 'eq', Subject.load(rule[0][-1])))
-                    for parent in potential_parents:
-                        q.append(Q('parents', 'eq', parent))
-            for sub in rule[0]:
-                q.append(Q('_id', 'eq', sub))
-        return Subject.find(reduce(lambda x, y: x | y, q)) if len(q) > 1 else (Subject.find(q[0]) if len(q) else Subject.find())
+        if self.subjects.exists():
```

Comment:
Added in https://github.com/CenterForOpenScience/osf.io/pull/7127/commits/a431d82ec0fe7ac1e8310766961ba86ea73ff9e5